File: HM/310_MinimumHeightTrees.py
from collections import defaultdict

def findMinHeightTrees(n: int, edges: List[List[int]]) -> List[int]:

    if n == 1:
        return [0]  # Only one node, so the root is the only solution

    # Build the graph
    graph = defaultdict(list)
    for edge in edges:
        u, v = edge
        graph[u].append(v)
        graph[v].append(u)

    #가장 많이 연결되어 있는 노드가 루트가 되어야 최소 높이 트리를 만들 수 있음. 순차적으로 리프노드와 가까운 노드들을 제거해나간다.

    #리프노드를 제거해나간다. 
    leaves = []
    for i in range(n+1):
        #해당 키의 value가 하나면 리프노드.
        if len(graph[i]) == 1:
            leaves.append(i)

    """
    n은 전체 노드의 개수이므로 여기서 leaves, 즉 리프 노드의 개수만큼 계속 빼나가면서 최종 2개 이하가 남을 때까지 반복한다. 마지막에 남은 값이 홀수 개일 때는 루트가 최종 1개가 되지만, 짝수 개일 때는 2개가 될 수 있다. 
    따라서 while 반복문은 2개까지는 계속 반복한다. 리프 노드는 반복하면서 제거한다. 그래프 딕셔너리에서 pop()으로 제거하고, 연결된 값도 찾아서 제거한다. 
    즉 무방향 그래프라 그래프를 각각 두 번씩 만들었으므로 제거 또한 두 번씩 진행한다. 이제 마찬가지로 값이 1개뿐일 때는 리프 노드라는 의미이므로, 새로운 리프 노드를 new_Leaves로 구성하여 교체한다.
    """
    #계속해서 연결작은 애들 없애기. 최종 2개 이하가 남을 때까지 반복 (홀수면 바로 루트지만 짝수일 수도 있기에)
    while n > 2:
        n = n - len(leaves)
        new_leaves = []
        for leaf in leaves:
            neighbor = graph[leaf].pop()
            graph[neighbor].remove(leaf)

            if len(graph[neighbor]) == 1:
                new_leaves.append(neighbor)

        leaves = new_leaves

    return leaves

File: HM/310_MinimumHeightTrees.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* findMinHeightTrees(int n, int** edges, int edgesSize, int* edgesColSize, int* returnSize) {
    if (n == 1) {
        int* result = (int*)malloc(sizeof(int));
        result[0] = 0;
        *returnSize = 1;
        return result;
    }

    int* graph[n];
    for (int i = 0; i < n; i++) {
        graph[i] = (int*)malloc(sizeof(int));
        graph[i][0] = 0;
    }

    // Build the graph
    for (int i = 0; i < edgesSize; i++) {
        int u = edges[i][0];
        int v = edges[i][1];

        graph[u] = (int*)realloc(graph[u], (graph[u][0] + 2) * sizeof(int));
        graph[v] = (int*)realloc(graph[v], (graph[v][0] + 2) * sizeof(int));

        graph[u][0]++;
        graph[v][0]++;

        graph[u][graph[u][0]] = v;
        graph[v][graph[v][0]] = u;
    }

    // Initialize leaves (nodes with only one connection)
    int* leaves = (int*)malloc(n * sizeof(int));
    int leavesCount = 0;

    for (int i = 0; i < n; i++) {
        if (graph[i][0] == 1) {
            leaves[leavesCount++] = i;
        }
    }

    // Remove leaves and update the graph until only 2 or fewer nodes are left
    while (n > 2) {
        n -= leavesCount;
        int* newLeaves = (int*)malloc(n * sizeof(int));
        int newLeavesCount = 0;

        for (int i = 0; i < leavesCount; i++) {
            int leaf = leaves[i];
            int neighbor = graph[leaf][1];

            // Remove the leaf from its neighbor's connections
            for (int j = 1; j <= graph[neighbor][0]; j++) {
                if (graph[neighbor][j] == leaf) {
                    for (int k = j; k < graph[neighbor][0]; k++) {
                        graph[neighbor][k] = graph[neighbor][k + 1];
                    }
                    graph[neighbor][0]--;
                    break;
                }
            }

            // If the neighbor becomes a leaf, add it to the new leaves array
            if (graph[neighbor][0] == 1) {
                newLeaves[newLeavesCount++] = neighbor;
            }
        }

        free(leaves);
        leaves = newLeaves;
        leavesCount = newLeavesCount;
    }

    *returnSize = leavesCount;
    return leaves;
}

